File: 11.PrefixSum前缀和/leetcode_3070_countSummatrices.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int countSubmatrices(vector<vector<int>>& grid, int k) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> preSum(m+1, vector<int>(n+1, 0));
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                preSum[i+1][j+1] = preSum[i+1][j]+preSum[i][j+1]+grid[i][j]-preSum[i][j];
            }
        }

        int res = 0;
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                if (preSum[i+1][j+1] > k){
                    break;
                }
                res++;
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<vector<int>> grid = {{7,6,3},{6,6,1}};
    int k = 18;
    cout << s.countSubmatrices(grid, k) << endl;
}

File: 11.PrefixSum前缀和/leetcode_3070_countSummatrices.py
from typing import List

class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        preSum = [[0]*(n+1) for _ in range(m+1)]
        for i in range(m):
            for j in range(n):
                preSum[i+1][j+1] = preSum[i+1][j]+preSum[i][j+1]+grid[i][j]-preSum[i][j]
        
        res =  0
        for i in range(m):
            for j in range(n):
                if preSum[i+1][j+1] > k:
                    break
                res += 1
        return res

if __name__ == "__main__":
    s = Solution()
    grid = [[7,6,3],[6,6,1]]
    k = 18
    print(s.countSubmatrices(grid, k))
